Libraries/LibWeb/IndexedDB/Internal/Algorithms.cpp

```diff
@@ -341,9 +349,149 @@ void upgrade_a_database(JS::Realm& realm, GC::Ref<IDBDatabase> connection, u64 v
+
+    // 11. Delete db. If this fails for any reason, return an appropriate error (e.g. "QuotaExceededError" or "UnknownError" DOMException).
+    auto maybe_deleted = Database::delete_for_key_and_name(storage_key, name);
+    if (maybe_deleted.is_error())
+        return WebIDL::OperationError::create(realm, "Unable to delete database"_string);
+
+    // 12. Return version.
+    return version;
+}
+
+// https://w3c.github.io/IndexedDB/#abort-a-transaction
+void abort_a_transaction(IDBTransaction& transaction, GC::Ptr<WebIDL::DOMException> error)
+{
+    // NOTE: This is not spec'ed anywhere, but we need to know IF the transaction was aborted.
+    transaction.set_aborted(true);
+
+    // FIXME: 1. All the changes made to the database by the transaction are reverted.
+    // For upgrade transactions this includes changes to the set of object stores and indexes, as well as the change to the version.
+    // Any object stores and indexes which were created during the transaction are now considered deleted for the purposes of other algorithms.
+
+    // FIXME: 2. If transaction is an upgrade transaction, run the steps to abort an upgrade transaction with transaction.
+    // if (transaction.is_upgrade_transaction())
+    //     abort_an_upgrade_transaction(transaction);
+
+    // 3. Set transaction’s state to finished.
+    transaction.set_state(IDBTransaction::TransactionState::Finished);
+
+    // 4. If error is not null, set transaction’s error to error.
+    if (error)
+        transaction.set_error(error);
+
+    // FIXME: 5. For each request of transaction’s request list, abort the steps to asynchronously execute a request for request,
+    //           set request’s processed flag to true, and queue a task to run these steps:
+    // FIXME: 1. Set request’s done flag to true.
+    // FIXME: 2. Set request’s result to undefined.
+    // FIXME: 3. Set request’s error to a newly created "AbortError" DOMException.
+    // FIXME: 4. Fire an event named error at request with its bubbles and cancelable attributes initialized to true.
+
+    // 6. Queue a task to run these steps:
+    HTML::queue_a_task(HTML::Task::Source::DatabaseAccess, nullptr, nullptr, GC::create_function(transaction.realm().vm().heap(), [&transaction]() {
+        // 1. If transaction is an upgrade transaction, then set transaction’s connection's associated database's upgrade transaction to null.
+        if (transaction.is_upgrade_transaction())
+            transaction.connection()->associated_database()->set_upgrade_transaction(nullptr);
+
+        // 2. Fire an event named abort at transaction with its bubbles attribute initialized to true.
+        transaction.dispatch_event(DOM::Event::create(transaction.realm(), HTML::EventNames::abort, { .bubbles = true }));
+
+        // 3. If transaction is an upgrade transaction, then:
+        if (transaction.is_upgrade_transaction()) {
+            // 1. Let request be the open request associated with transaction.
```

Comment:
It sorta-kinda does, in a very indirect way.. (Like all of this spec 🙃)
`The transaction of an open request is null unless an upgradeneeded event has been fired.`

Reply:
The `open request associated with transaction` is not the same as `transaction of an open request`, it only seems to imply that this is the case... similar to how a request can be associated to a transaction through the latter's "request list".

Reply:
Yes. Thats my main issue implementing this. A ton of stuff is implied.. 
`the open request associated with transaction` is simple enough, since its exclusive.

See the note here: https://w3c.github.io/IndexedDB/#upgrade-transaction-construct

Reply:
Right, I agree with your implementation, just wondering whether this is something to file an issue over :-)

Reply:
Theres a few other things ive found that also warrants a spec issue, so i can go over the ad-hoc places and make a spec issue for them :)

Reply:
I think this should cover all of it: https://github.com/w3c/IndexedDB/issues/433, but im sure more will show up as it progresses